libcxx/test/std/utilities/function.objects/func.memfn/mem_fn.pass.cpp

```diff
@@ -0,0 +1,741 @@
+  f(a) = 5;
+  assert(a.data_ == 5);
+  A* ap = &a;
+  f(ap) = 6;
+  assert(a.data_ == 6);
+  const A* cap = ap;
+  assert(f(cap) == f(ap));
+  const F& cf = f;
+  assert(cf(ap) == f(ap));
+
+#if TEST_STD_VER >= 11
+  static_assert(noexcept(f(a)), "");
+  static_assert(noexcept(f(ap)), "");
+  static_assert(noexcept(f(cap)), "");
+  static_assert(noexcept(cf(ap)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_fun0(F f) {
+  A a = {};
+  assert(f(a) == 'a');
+  A* ap = &a;
+  assert(f(ap) == 'a');
+  const F& cf = f;
+  assert(cf(ap) == 'a');
+
+#if TEST_STD_VER >= 17
+  static_assert(!noexcept(f(a)), "");
+  static_assert(!noexcept(f(ap)), "");
+  static_assert(!noexcept(cf(ap)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_fun1(F f) {
+  A a = {};
+  assert(f(a, 1) == 'b');
+  A* ap = &a;
+  assert(f(ap, 2) == 'b');
+  const F& cf = f;
+  assert(cf(ap, 2) == 'b');
+
+#if TEST_STD_VER >= 17
+  static_assert(!noexcept(f(a, 0)), "");
+  static_assert(!noexcept(f(ap, 1)), "");
+  static_assert(!noexcept(cf(ap, 2)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_fun2(F f) {
+  A a = {};
+  assert(f(a, 1, 2) == 'c');
+  A* ap = &a;
+  assert(f(ap, 2, 3.5) == 'c');
+  const F& cf = f;
+  assert(cf(ap, 2, 3.5) == 'c');
+
+#if TEST_STD_VER >= 17
+  static_assert(!noexcept(f(a, 0, 0.0)), "");
+  static_assert(!noexcept(f(ap, 1, 2)), "");
+  static_assert(!noexcept(cf(ap, 2, 3.5)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_noexcept_fun0(F f) {
+  A a = {};
+  assert(f(a) == 'd');
+  A* ap = &a;
+  assert(f(ap) == 'd');
+  const F& cf = f;
+  assert(cf(ap) == 'd');
+
+#if TEST_STD_VER >= 17
+  static_assert(noexcept(f(a)), "");
+  static_assert(noexcept(f(ap)), "");
+  static_assert(noexcept(cf(ap)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_noexcept_fun1(F f) {
+  A a = {};
+  assert(f(a, 1) == 'e');
+  A* ap = &a;
+  assert(f(ap, 2) == 'e');
+  const F& cf = f;
+  assert(cf(ap, 2) == 'e');
+
+#if TEST_STD_VER >= 17
+  static_assert(noexcept(f(a, 0)), "");
+  static_assert(noexcept(f(ap, 1)), "");
+  static_assert(noexcept(cf(ap, 2)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_noexcept_fun2(F f) {
+  A a = {};
+  assert(f(a, 1, 2) == 'f');
+  A* ap = &a;
+  assert(f(ap, 2, 3.5) == 'f');
+  const F& cf = f;
+  assert(cf(ap, 2, 3.5) == 'f');
+
+#if TEST_STD_VER >= 17
+  static_assert(noexcept(f(a, 0, 0.0)), "");
+  static_assert(noexcept(f(ap, 1, 2)), "");
+  static_assert(noexcept(cf(ap, 2, 3.5)), "");
+#endif
+
+  return true;
+}
+
+template <class F>
+TEST_CONSTEXPR_CXX20 bool test_const_fun0(F f) {
+  A a = {};
+  assert(f(a) == 'a');
+  A* ap = &a;
+  assert(f(ap) == 'a');
+  const A* cap = &a;
+  assert(f(cap) == 'a');
+  const F& cf = f;
+  assert(cf(ap) == 'a');
+
+#if TEST_STD_VER >= 17
```

Comment:
Why are you only checking for noexcept-ness in >= C++17?

Reply:
In earlier modes, the stored member function pointer couldn't propagate noexcept-ness, because noexcept-specifier is made a part of type only since C++17 via [P0012R1](https://wg21.link/p0012r1).

I'm not sure whether the behavioral change of the C++ core language should be reflected here.